src/ema/math.rs

```diff
@@ -0,0 +1,213 @@
+    debug_assert!(smoothing <= FixedU128::one());
+    let complement = FixedU128::one() - smoothing;
+    // in order to determine the iterated smoothing factor we exponentiate the complement
+    let exp_complement = complement.saturating_pow(iterations as usize);
+    debug_assert!(exp_complement <= FixedU128::one());
+    FixedU128::one() - exp_complement
+}
+
+/// Calculates smoothing factor alpha for an exponential moving average based on `period`:
+/// `alpha = 2 / (period + 1)`. It leads to the "center of mass" of the EMA corresponding to be the
+/// "center of mass" of a `period`-length SMA.
+///
+/// Possible alternatives for `alpha = 2 / (period + 1)`:
+/// + `alpha = 1 - 0.5^(1 / period)` for a half-life of `period` or
+/// + `alpha = 1 - 0.5^(2 / period)` to have the same median as a `period`-length SMA.
+/// See https://en.wikipedia.org/wiki/Moving_average#Relationship_between_SMA_and_EMA
+pub fn smoothing_from_period(period: u64) -> FixedU128 {
+    FixedU128::saturating_from_rational(2u64, period.max(1).saturating_add(1))
+}
+
+/// Calculate a weighted average for the given prices.
+/// `prev` is the previous oracle value, `incoming` is the new value to integrate.
+/// `weight` is how much weight to give the new value.
+///
+/// Note: Rounding is slightly biased towards `prev`.
+/// (`FixedU128::mul` rounds to the nearest representable value, rounding down on equidistance.
+/// See [doc comment here](https://github.com/paritytech/substrate/blob/ce10b9f29353e89fc3e59d447041bb29622def3f/primitives/arithmetic/src/fixed_point.rs#L670-L671).)
+pub fn price_weighted_average(prev: Price, incoming: Price, weight: FixedU128) -> Price {
+    debug_assert!(weight <= FixedU128::one(), "weight must be <= 1");
+    if incoming >= prev {
+        // Safe to use bare `+` because `weight <= 1` and `a + (b - a) <= b`.
+        // Safe to use bare `-` because of the conditional.
+        prev + weight * (incoming - prev)
+    } else {
+        // Safe to use bare `-` because `weight <= 1` and `a - (a - b) >= 0` and the conditional.
+        prev - weight * (prev - incoming)
+    }
+}
+
+/// Calculate a weighted average for the given balances.
+/// `prev` is the previous oracle value, `incoming` is the new value to integrate.
+/// `weight` is how much weight to give the new value.
+///
+/// Note: Rounding is biased towards `prev`.
+pub fn balance_weighted_average(prev: Balance, incoming: Balance, weight: FixedU128) -> Balance {
+    debug_assert!(weight <= FixedU128::one(), "weight must be <= 1");
+    if incoming >= prev {
+        // Safe to use bare `+` because `weight <= 1` and `a + (b - a) <= b`.
+        // Safe to use bare `-` because of the conditional.
+        prev + weight.saturating_mul_int(incoming - prev)
+    } else {
+        // Safe to use bare `-` because `weight <= 1` and `a - (a - b) >= 0` and the conditional.
+        prev - weight.saturating_mul_int(prev - incoming)
```

Comment:
Same as previous comment would apply here and everywhere where "unsafe" math is used, although safety is/should be guaranteed by check conditions.